File: reflect_demo/src/main/java/com/example/reflect_demo/MainActivity.java
package com.example.reflect_demo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;

public class MainActivity extends AppCompatActivity {

    private TextView tvShowStaff;
    public static final String TAG = MainActivity.class.getSimpleName();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvShowStaff = ((TextView) findViewById(R.id.tvShowStaff));
        reflect();
    }

    private void reflect() {
        //（1）方法一：
        Staff staff = new Staff();
        Class<? extends Staff> aClass = staff.getClass();
        Log.d(TAG, "reflect: "+aClass.getSimpleName());
        //（2）方法二：
        Class aClass1 = Staff.class;
        Log.d(TAG, "reflect: "+aClass1.getSimpleName()+"     "+aClass1);
        //  方法3：Class.forName()
        try {
            Class<?> aClass2 = Class.forName("com.example.reflect_demo.Staff");
            int modifiers = aClass2.getModifiers();
            String s = Modifier.toString(modifiers);
            Log.d(TAG, "reflect: 修饰符    "+s);
            Field[] fields = aClass2.getFields();
            String s1 = Arrays.toString(fields);
            Log.d(TAG, "reflect: 可用字段  "+s1);
            Constructor<?>[] constructors = aClass2.getConstructors();
            Log.d(TAG, "reflect: 构造方法  "+Arrays.toString(constructors));
            Method declaredMethod = aClass2.getDeclaredMethod("work");
            Log.d(TAG, "reflect: 声明方法   "+declaredMethod.getName());
            Staff staff1 = (Staff) aClass2.newInstance();
            staff1.setAge(20);
            staff1.setName("ls");
            staff1.setDepart("mobile development");
            Object invoke = declaredMethod.invoke(staff1);
            Log.d(TAG, "reflect: 反射方法已执行  ");

            Object o = aClass1.newInstance();
            Object invoke1 = declaredMethod.invoke(o);

        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        }
    }
}

File: build_pattern/src/main/java/com/example/build_pattern/MainActivity.kt
package com.example.build_pattern

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import com.example.build_pattern.build.ConcreteBuilder
import com.example.build_pattern.build.Director
import kotlinx.android.synthetic.main.activity_main.*

/**
 * 建造者模式属于创建型模式。
建造者模式主要用来创建复杂的对象，用户可以不用关心其建造过程和细节。
例如：当要组装一台电脑时，我们选择好CPU、内存、硬盘等等，然后交给装机师傅，装机师傅就把电脑给组装起来，我们不需要关心是怎么拼装起来的。
 */
class MainActivity : AppCompatActivity() {

    val TAG = ConcreteBuilder::class.java.simpleName

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        StringUtils.setPricePoint(et)


    }
    public fun createComputer(view: View){

        var toString = et.text.toString()
        Log.d(TAG, "当前票价    $toString")
        var builder = ConcreteBuilder()//创建建造者实例，（装机人员）
        var direcror =  Director(builder)//创建指挥者实例，并分配相应的建造者，（老板分配任务）
        direcror.Construct("i7-6700", "三星DDR4", "希捷1T")//组装电脑
        var computer = builder.create()
        Log.e(TAG,computer.toString())

    }
}
